river/rule_list.zig

```diff
@@ -0,0 +1,129 @@
+        rules: std.ArrayListUnmanaged(Rule) = .{},
+
+        pub fn deinit(list: *Self) void {
+            for (list.rules.items) |rule| {
+                util.gpa.free(rule.app_id_glob);
+                util.gpa.free(rule.title_glob);
+            }
+            list.rules.deinit(util.gpa);
+        }
+
+        pub fn add(list: *Self, rule: Rule) error{OutOfMemory}!void {
+            const index = for (list.rules.items) |*existing, i| {
+                if (mem.eql(u8, rule.app_id_glob, existing.app_id_glob) and
+                    mem.eql(u8, rule.title_glob, existing.title_glob))
+                {
+                    existing.value = rule.value;
+                    return;
+                }
+
+                switch (globber.order(rule.app_id_glob, existing.app_id_glob)) {
+                    .lt => break i,
+                    .eq => {
+                        if (globber.order(rule.title_glob, existing.title_glob) == .lt) {
+                            break i;
+                        }
+                    },
+                    .gt => {},
+                }
+            } else list.rules.items.len;
+
+            const owned_app_id_glob = try util.gpa.dupe(u8, rule.app_id_glob);
+            errdefer util.gpa.free(owned_app_id_glob);
+
+            const owned_title_glob = try util.gpa.dupe(u8, rule.title_glob);
+            errdefer util.gpa.free(owned_title_glob);
+
+            try list.rules.insert(util.gpa, index, .{
+                .app_id_glob = owned_app_id_glob,
+                .title_glob = owned_title_glob,
+                .value = rule.value,
+            });
+        }
+
+        pub fn del(list: *Self, rule: Rule) void {
+            for (list.rules.items) |existing, i| {
+                if (mem.eql(u8, rule.app_id_glob, existing.app_id_glob) and
+                    mem.eql(u8, rule.title_glob, existing.title_glob))
+                {
+                    util.gpa.free(existing.app_id_glob);
+                    util.gpa.free(existing.title_glob);
+                    _ = list.rules.orderedRemove(i);
+                    return;
+                }
+            }
+        }
+
+        /// Returns the value of the most specific rule matching the view.
+        /// Returns null if no rule matches.
+        pub fn match(list: *Self, view: *View) ?T {
+            const app_id = mem.sliceTo(view.getAppId(), 0) orelse "";
+            const title = mem.sliceTo(view.getTitle(), 0) orelse "";
+
+            for (list.rules.items) |rule| {
+                if (globber.match(app_id, rule.app_id_glob) and
+                    globber.match(title, rule.title_glob))
+                {
+                    return rule.value;
+                }
+            }
+
+            return null;
+        }
+
+        /// Returns the length of the longest globs.
+        pub fn getMaxGlobLen(self: *const Self) MaxGlobLen {
+            var app_id_len: usize = "*".len;
+            var title_len: usize = "*".len;
```

Comment:
These initial values aren't necessary for correctness so I think it is less confusing to use 0 here. Slices of length 0 are rejected by the glob validation function.
```suggestion
            var app_id_len: usize = 0;
            var title_len: usize = 0;
```